### ISSUE TITLE
Feature request: counsel-recentf with files ordered by last visible

### ISSUE THREAD

**Issue by wolfv6** [abo-abo/swiper#1560]
Currently counsel-recentf lists file history in minibuffer in the order they where opened.
- if an opened file is not one of the 12 most recent files opened, it is place at the top of history
- if an opened file is one of the 12 most recent files opened, it's place in history remains unchanged

There is a more human friendly way to list file history.
Instead of listing files in the order they were opened, list files in the order they were last visible.

This is more human friendly because it is how human memory works.
I remember when I was last looking at something (not when I opened it).
So when I think "I was just looking at such-and-such.  What file was that in?", I find the file listed near the top of the minibuffer.

Here is a more detailed description:
List currently visible files at the top of minibuffer.
Below currently visible files, list file history in the order that they where last visible.
Every time a file disappears from view, it is moved to just below currently visible files.

Use cases:
- reference a file often, while editing multiple files in another window
- edit a file often, while referencing multiple files in another window

Example minibuffer counsel-recentflv with 2 windows currently open:
```
file currently open in window1 (currently open files are ordered as they appear in the frame)
file currently open in window2
file last visible at 3:10 (not visible files are ordered by times they where last visible)
file last visible at 3:15
file last visible at 3:19
file last visible at 3:30
```

**Comment by abo-abo** [comment]
This sounds more like a feature request for `recentf`. `counsel-recentf` is simply a client of `recentf`, it should not take over the responsibilities of that package.

So I suggest to open a feature request for `recentf` with <kbd>M-x</kbd> `report-emacs-bug`, since `recentf` is in Emacs core. Once the functionality is implemented there, `counsel-recentf` can use it.

If getting things done in the Emacs core seems too slow, you could also set up a new package on MELPA, e.g. `better-recentf` or something.

**Comment by basil-conto** [comment]
For posterity, this has been submitted as [bug#31385](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=31385).

**Comment by zzamboni** [comment]
@basil-conto thanks for submitthing this. Did you ever see any progress in that recentf bug, or find a workaround? I have been wondering about this behavior myself recently.

**Comment by basil-conto** [comment]
@zzamboni No, sorry. I have not followed up on it because I do not use `recentf`, but as you can see in the linked feature request, neither has anyone else.

**Comment by tau3000** [comment]
@wolfv6 @zzamboni The following works well for me.
```
(use-package switch-buffer-functions
  :defer t
  :after recentf
  :preface
  (defun my-recentf-track-visited-file (_prev _curr)
    (and buffer-file-name
         (recentf-add-file buffer-file-name)))
  :init
  (add-hook 'switch-buffer-functions #'my-recentf-track-visited-file))
```

cf. https://tsuu32.hatenablog.com/entry/2019/12/16/124052 (in Japanese)

**Comment by zzamboni** [comment]
@tau3000 thanks for the tip! Works well for me too. I've added this to my config: https://github.com/zzamboni/dot-doom/blob/master/doom.org#miscellaneous-keybindings
